components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/AccountStorage.kt

```diff
@@ -48,3 +73,75 @@ class SharedPrefAccountStorage(val context: Context) : AccountStorage {
         return context.getSharedPreferences(FXA_STATE_PREFS_KEY, Context.MODE_PRIVATE)
     }
 }
+
+/**
+ * A base class for exceptions describing abnormal account storage behaviour.
+ */
+internal abstract class AbnormalAccountStorageEvent : Exception() {
+    /**
+     * Account state was expected to be present, but it wasn't.
+     */
+    internal class UnexpectedlyMissingAccountState : AbnormalAccountStorageEvent()
+}
+
+/**
+ * Account storage layer which uses encrypted-at-rest storage implementation for supported API levels (23+).
+ * On older API versions account state is stored in plaintext.
+ *
+ * Migration from [SharedPrefAccountStorage] will happen upon initialization,
+ * unless disabled via [migrateFromPlaintextStorage].
+ */
+internal class SecureAbove22AccountStorage(
+    context: Context,
+    private val crashReporter: CrashReporter? = null,
+    migrateFromPlaintextStorage: Boolean = true
+) : AccountStorage {
+    companion object {
+        private const val STORAGE_NAME = "fxaStateAC"
+        private const val KEY_ACCOUNT_STATE = "fxaState"
+        private const val PREF_NAME = "fxaStatePrefAC"
+        private const val PREF_KEY_HAS_STATE = "fxaStatePresent"
+    }
+
+    private val store = SecureAbove22Preferences(context, STORAGE_NAME)
+    // Prefs are used here to keep track of abnormal storage behaviour - namely, account state disappearing without
+    // being cleared first through this class. Note that clearing application data will clear both 'store' and 'prefs'.
+    private val prefs = context.getSharedPreferences(PREF_NAME, Context.MODE_PRIVATE)
+
+    init {
+        if (migrateFromPlaintextStorage) {
+            // In case we switched from SharedPrefAccountStorage to this implementation, migrate persisted account
+            // and clear out the old storage layer.
+            val plaintextStorage = SharedPrefAccountStorage(context, migrateFromSecureStorage = false)
```

Comment:
Not sure where my previous comment went. It was just a nit to ask if we could instead add `migrateFrom...` methods to the storages. It seems the constructor parameter is easy to miss and then we'd migrate by mistake. With a method we'd trigger migration explicitly.

Reply:
With an explicit method, the only way to get rid off the constructor flag is to expose a `migrate` method as part of `AccountStorage` interface; i really like the enclosed nature of these classes - consumers don't need to know anything about migrations, it "just works". This is how the lower level works, as well - consumers of secure prefs don't need to know that it auto-migrates data from insecure prefs in case of an API upgrade! So there's a nice symmetry here I'd like to keep. So, landing as-is.